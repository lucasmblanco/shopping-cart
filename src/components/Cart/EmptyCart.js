import React from 'react'
import {HiOutlineEmojiSad } from 'react-icons/hi'

export default function EmptyCart() {
  return (
    <div className='flex flex-col items-center justify-center text-lightest-blue-accent h-full text-2xl '>
                            <HiOutlineEmojiSad />
                            <div className='text-lg p-2'>Oops! Your cart is empty!</div>
    </div>
  )
}
