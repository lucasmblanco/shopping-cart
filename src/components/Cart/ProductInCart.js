import React, { useEffect, useState } from 'react';
import {HiTrash} from 'react-icons/hi'

export default function ProductInCart({product, deleteProduct, addQuantity}) {

  //  const [ quantity, dispatch ] = useReducer(reducer, product.quantity); 
  //  const [price, dispatch2] = useReducer(reducer, product.price); 
    const [ limit, setLimit ] = useState(false)
    
    useEffect(() => {
        product.quantity <= 1 ? setLimit(true) : setLimit(false)
      }, 
      [product.quantity])

  return (
   <>
    <div className=' flex justify-around text-xs text-left text-white-accent h-16 '>
        <div className='grid content-center text-[8px] w-[10rem]  text-white-accent  p-4'>{product.title}</div>
        <div className='w-[3rem] grid grid-cols-cart-quantity bg-blue-accent m-1 rounded-md'> 
            <div className='col-span-2 justify-self-center self-end'>{product.quantity}</div>
            <button className='text-green-500 transition-all duration-500 hover:text-green-300'onClick={()=> {addQuantity(product.id, 'decrement')}} disabled={limit}>-</button>
            <button className='text-red-500 transition-all duration-500 hover:text-red-300' onClick={()=> {addQuantity(product.id, 'increment')}}>+</button>
     </div>
        <div className=' grid content-center justify-center text-base  w-[5rem] transition-all duration-500'>$ {product.price * product.quantity}</div>
        <div className="flex justify-center rounded-full text-red-accent transition-all duration-500 hover:text-light-red-accent p-4"><button onClick={() => deleteProduct(product.id)}><HiTrash /></button></div>
    </div>
    <hr></hr>
   </>
  )
}
