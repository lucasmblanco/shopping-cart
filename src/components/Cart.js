import React, { useState } from 'react';
import ProductInCart from './Cart-components/ProductInCart';
import { HiShoppingCart } from 'react-icons/hi';
import EmptyCart from './Cart-components/EmptyCart';

export default function Cart({ products, deleteProduct, addQuantity, total }) {
  const [open, setOpen] = useState(false);

  function openCart() {
    setOpen(!open);
  }

  return (
    <>
      <button
        data-testid="open"
        className=" h-12 text-white font-inter rounded-tr-xl rounded-br-xl bg-red-accent border-light-red-accent 
        border-t-2 border-b-2 border-r-2 p-4"
        onClick={openCart}>
        <div className="hover:drop-shadow-nxl transition-all duration-500 ">
          <HiShoppingCart />
        </div>
      </button>
      {products.length > 0 && (
        <div
          className="flex items-center justify-center text-sm absolute 
        left-[64.5rem] top-1 rounded-full h-5 w-5 animate-fade-in-bck-open bg-lightest-blue-accent font-inter bg-opacity-75 ">
          {products.reduce((acc, product) => acc + product.quantity, 0)}
        </div>
      )}

      {open && (
        <div
          data-testid="cart-body"
          className="flex flex-col absolute m-3 w-96 h-96 top-[3rem] desktop:top-[3rem] left-[20rem] desktop:left-cart-left-position p-4 text-black 
                rounded-[calc(0.5rem_+_2px)] backdrop-blur-3xl backdrop-brightness-50
                font-inter text-center animate-open-anim">
          <div className="flex justify-center p-2 text-lightest-blue-accent">
            <HiShoppingCart />
          </div>

          <div className="overflow-y-auto h-80 ">
            {products.length ? (
              products.map((product) => {
                return (
                  <ProductInCart
                    key={product.id}
                    product={product}
                    deleteProduct={deleteProduct}
                    addQuantity={addQuantity}
                  />
                );
              })
            ) : (
              <EmptyCart />
            )}
          </div>

          <div className="flex p-2 justify-between bg-lightest-blue-accent rounded-xl m-2">
            <div className="flex items-center p-2">Total: ${total}</div>
            <button
              className="text-white-accent p-2 transition-all duration-500 
                        bg-red-accent rounded-lg hover:bg-light-red-accent border-2 border-light-red-accent hover:border-red-accent ">
              Go to Checkout
            </button>
          </div>
        </div>
      )}
    </>
  );
}
