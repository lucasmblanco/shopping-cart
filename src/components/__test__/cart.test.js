import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import Cart from '../Cart';

test('should render component', () => {
  const { asFragment } = render(<Cart products={[]} />);

  expect(asFragment()).toMatchSnapshot();
});

test('should render a button component', () => {
  render(<Cart products={[]} />);

  expect(screen.getByTestId('open')).toBeInTheDocument();
});

it('should open the cart body and must be empty', async () => {
  render(<Cart products={[]} />);

  const button = screen.getByTestId('open');

  userEvent.click(button);

  expect(await screen.findByTestId('cart-body')).toBeInTheDocument();
  expect(await screen.findByTestId('Empty-cart')).toBeInTheDocument();
});

test('should open the cart body and with a product', async () => {
  render(<Cart products={[{ id: 1, title: 'Product', quantity: 1 }]} />);

  const button = screen.getByTestId('open');

  await userEvent.click(button);

  expect(await screen.findByTestId('product-in-cart')).toBeInTheDocument();
});
