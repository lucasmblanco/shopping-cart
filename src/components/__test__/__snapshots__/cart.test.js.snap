// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component 1`] = `
<DocumentFragment>
  <button
    class=" h-12 text-white font-inter rounded-tr-xl rounded-br-xl bg-red-accent border-light-red-accent 
 border-t-2 border-b-2 border-r-2 p-4"
    data-testid="open"
  >
    <div
      class="hover:drop-shadow-nxl transition-all duration-500 "
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 1a1 1 0 000 2h1.22l.305 1.222a.997.997 0 00.01.042l1.358 5.43-.893.892C3.74 11.846 4.632 14 6.414 14H15a1 1 0 000-2H6.414l1-1H14a1 1 0 00.894-.553l3-6A1 1 0 0017 3H6.28l-.31-1.243A1 1 0 005 1H3zM16 16.5a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0zM6.5 18a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
        />
      </svg>
    </div>
  </button>
</DocumentFragment>
`;
